framework/src/meshdivisions/SubdomainsDivision.C

```diff
@@ -19,35 +19,55 @@ InputParameters
 SubdomainsDivision::validParams()
 {
   InputParameters params = MeshDivision::validParams();
+  params += BlockRestrictable::validParams();
   params.addClassDescription(
       "Divide the mesh by increasing subdomain ids. The division will be contiguously "
       "numbered even if the subdomain ids are not");
   return params;
 }
 
 SubdomainsDivision::SubdomainsDivision(const InputParameters & parameters)
-  : MeshDivision(parameters)
+  : MeshDivision(parameters), BlockRestrictable(this)
 {
   SubdomainsDivision::initialize();
 }
 
 void
 SubdomainsDivision::initialize()
 {
-  // The subdomains may not be contiguously numbered
-  std::set<libMesh::subdomain_id_type> subdomain_ids;
-  _mesh.getMesh().subdomain_ids(subdomain_ids, /*global=*/true);
-  setNumDivisions(subdomain_ids.size());
+  if (!blockRestricted())
+  {
+    // The subdomains may not be contiguously numbered
+    std::set<libMesh::subdomain_id_type> subdomain_ids;
+    _mesh.getMesh().subdomain_ids(subdomain_ids, /*global=*/true);
 
-  unsigned int i = 0;
-  for (const auto sub_id : subdomain_ids)
-    _subdomain_ids_to_division_index[sub_id] = i++;
+    setNumDivisions(subdomain_ids.size());
+    unsigned int i = 0;
+    for (const auto sub_id : subdomain_ids)
+      _subdomain_ids_to_division_index[sub_id] = i++;
+  }
+  else
+  {
+    // Follow the user input order, use a vector instead of a set
+    const auto subdomain_ids = _mesh.getSubdomainIDs(getParam<std::vector<SubdomainName>>("block"));
```

Comment:
Could call `getSubdomainIDs` with `blocks()`

Could also create an API on `BlockRestrictable` that returns its `_vec_ids`